File: DSA/linked_list/reverse_list.py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None

a = Node('A')
b = Node('B')
c = Node('C')
d = Node('D')

a.next = b
b.next = c
c.next = d

def print_linked_list(head):
    current = head
    while current:
        print(current.val)
        current = current.next

def reverse_list(node):
    pre = None

    while node:
        nex = node.next
        node.next = pre
        pre = node
        node = nex
    return pre

def reverse_list_rec(node, prev=None):
    if node == None:
        return pre
    nex = node.next
    node.next = prev
    prev = node
    return reverse_list_rec(nex, prev)

if __name__ == '__main__':
    print('Before reversing')
    print_linked_list(a)
    print('After reversing')
    print_linked_list(reverse_list(a))    
    print('After reversing recursively')
    print_linked_list(reverse_list(d))

File: DSA/data_structures/linked_list/reverseList.js
class Node {
	constructor(val) {
		this.val = val
		this.next = null
	}
}

const a = new Node('A')
const b = new Node('B')
const c = new Node('C')
const d = new Node('D')

a.next = b
b.next = c
c.next = d

const reverseList = (node) => {
	let pre = null

	while (node != null) {
		let nex = node.next
		node.next = pre
		pre = node
		node = nex
	}
	return pre
}

const printLinkedList = (head) => {
	let current = head
	while (current != null) {
		console.log(current.val)
		current = current.next
	}
}

const reverseListRec = (node, pre=null) => {
	if (node === null) return pre
	nex = node.next
	node.next = pre
	pre = node
	return reverseListRec(nex, pre)
}
	
console.log('Before reversing')
printLinkedList(a)
console.log('After reversing')
printLinkedList(reverseList(a))
console.log('After reversing recursively')
printLinkedList(reverseListRec(d))
